example/advanced/server-io/advanced_server_io.cpp

```diff
@@ -0,0 +1,914 @@
+    tcp::socket socket_;
+    tcp::acceptor acceptor_;
+    std::shared_ptr<std::string const> doc_root_;
+
+public:
+    listener(
+        round_robin& io_pool,
+        tcp::endpoint endpoint,
+        std::shared_ptr<std::string const> const& doc_root)
+        : io_pool_(io_pool)
+        , socket_(io_pool.get())
+        , acceptor_(io_pool.get())
+        , doc_root_(doc_root)
+    {
+        boost::system::error_code ec;
+
+        // Open the acceptor
+        acceptor_.open(endpoint.protocol(), ec);
+        if(ec)
+        {
+            fail(ec, "open");
+            return;
+        }
+
+        // Allow address reuse
+        acceptor_.set_option(boost::asio::socket_base::reuse_address(true), ec);
+        if(ec)
+        {
+            fail(ec, "set_option");
+            return;
+        }
+
+        // Bind to the server address
+        acceptor_.bind(endpoint, ec);
+        if(ec)
+        {
+            fail(ec, "bind");
+            return;
+        }
+
+        // Start listening for connections
+        acceptor_.listen(
+            boost::asio::socket_base::max_listen_connections, ec);
+        if(ec)
+        {
+            fail(ec, "listen");
+            return;
+        }
+    }
+
+    // Start accepting incoming connections
+    void
+    run()
+    {
+        if(! acceptor_.is_open())
+            return;
+        do_accept();
+    }
+
+    void
+    do_accept()
+    {
+        acceptor_.async_accept(
+            socket_,
+            std::bind(
+                &listener::on_accept,
+                shared_from_this(),
+                std::placeholders::_1));
+    }
+
+    void
+    on_accept(boost::system::error_code ec)
+    {
+        if(ec)
+        {
+            fail(ec, "accept");
+        }
+        else
+        {
+            // Create the http_session and run it
+            std::make_shared<http_session>(
+                std::move(socket_),
+                doc_root_)->run();
+        }
+
+        // reset socket with next io_context from the pool
+        socket_ = std::move(tcp::socket(io_pool_.get()));
+
+        // Accept another connection
+        do_accept();
+    }
+};
+
+//------------------------------------------------------------------------------
+
+int main(int argc, char* argv[])
+{
+    // Check command line arguments.
+    if (argc != 5)
+    {
+        std::cerr <<
+            "Usage: advanced-server <address> <port> <doc_root> <threads>\n" <<
+            "Example:\n" <<
+            "    advanced-server 0.0.0.0 8080 . 1\n";
+        return EXIT_FAILURE;
+    }
+    auto const address = boost::asio::ip::make_address(argv[1]);
+    auto const port = static_cast<unsigned short>(std::atoi(argv[2]));
+    auto const doc_root = std::make_shared<std::string>(argv[3]);
+    auto const threads = std::max<std::size_t>(1, std::atoi(argv[4]));
+
+    // The io_context is required for all I/O
+    // Use one io_context per thread round robin style
+    round_robin io_pool {threads};
```

Comment:
I think this would look better like this:
```
multi_io_context ioc{threads};
```

Reply:
Sure